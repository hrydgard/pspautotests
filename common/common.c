#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <assert.h>
#include <pspdebug.h>
#include <pspthreadman.h>
#include <psploadexec.h>
#include <pspctrl.h>
#include <pspdisplay.h>
#include <pspiofilemgr.h>
//#include <pspkdebug.h>

#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <sys/lock.h>
#include <sys/fcntl.h>

//#include "local.h"

#include "sysmem-imports.h"

/*
enum PspThreadAttributes
{
	PSP_THREAD_ATTR_VFPU = 0x00004000,
	PSP_THREAD_ATTR_USER = 0x80000000,
	PSP_THREAD_ATTR_USBWLAN = 0xa0000000,
	PSP_THREAD_ATTR_VSH = 0xc0000000,
	PSP_THREAD_ATTR_SCRATCH_SRAM = 0x00008000,
	PSP_THREAD_ATTR_NO_FILLSTACK = 0x00100000,
	PSP_THREAD_ATTR_CLEAR_STACK = 0x00200000,
};
*/
/*
enum PspModuleInfoAttr
{
	PSP_MODULE_USER			= 0,
	PSP_MODULE_NO_STOP		= 0x0001,
	PSP_MODULE_SINGLE_LOAD	= 0x0002,
	PSP_MODULE_SINGLE_START	= 0x0004,
	PSP_MODULE_KERNEL		= 0x1000,
};
*/

PSP_MODULE_INFO("TESTMODULE", PSP_MODULE_USER, 1, 0);
PSP_MAIN_THREAD_ATTR(PSP_THREAD_ATTR_USER | PSP_THREAD_ATTR_VFPU);
//PSP_MAIN_THREAD_ATTR(PSP_THREAD_ATTR_USER);

#define EMULATOR_DEVCTL__GET_HAS_DISPLAY 0x00000001
#define EMULATOR_DEVCTL__SEND_OUTPUT     0x00000002
#define EMULATOR_DEVCTL__IS_EMULATOR     0x00000003
#define EMULATOR_DEVCTL__SEND_CTRLDATA   0x00000010
#define EMULATOR_DEVCTL__EMIT_SCREENSHOT 0x00000020

unsigned int RUNNING_ON_EMULATOR = 0;
unsigned int CHECKPOINT_ENABLE_TIME = 0;
unsigned int CHECKPOINT_OUTPUT_DIRECT = 0;
unsigned int HAS_DISPLAY = 1;

// 21 MB to give space for thread stacks and etc.
unsigned int sce_newlib_heap_kb_size = 21504;

extern int test_main(int argc, char *argv[]);

FILE stdout_back = {NULL};
//int KprintfFd = 0;

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	if (CHECKPOINT_OUTPUT_DIRECT) {
		// This is easier to debug in the emulator, but printf() reschedules on the real PSP.
		vprintf(format, args);
	} else {
		schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	}
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

SceUID reschedThread;
volatile int didResched = 0;
int reschedFunc(SceSize argc, void *argp) {
	didResched = 1;
	return 0;
}

u64 lastCheckpoint = 0;
void checkpoint(const char *format, ...) {
	u64 currentCheckpoint = sceKernelGetSystemTimeWide();
	if (CHECKPOINT_ENABLE_TIME) {
		schedf("[%s/%lld] ", didResched ? "r" : "x", currentCheckpoint - lastCheckpoint);
	} else {
		schedf("[%s] ", didResched ? "r" : "x");
	}

	sceKernelTerminateThread(reschedThread);

	if (format != NULL) {
		va_list args;
		va_start(args, format);
		if (CHECKPOINT_OUTPUT_DIRECT) {
			// This is easier to debug in the emulator, but printf() reschedules on the real PSP.
			vprintf(format, args);
		} else {
			schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
		}
		va_end(args);
	}

	didResched = 0;
	sceKernelStartThread(reschedThread, 0, NULL);

	if (format != NULL) {
		schedf("\n");
	}
	
	lastCheckpoint = currentCheckpoint;
}

void checkpointNext(const char *title) {
	if (schedfBufferPos != 0) {
		schedf("\n");
	}
	flushschedf();
	didResched = 0;
	if (title != NULL)
		checkpoint(title);
}

static int writeStdoutHook(struct _reent *ptr, void *cookie, const char *buf, int buf_len) {
	char temp[1024 + 1];

	//if (KprintfFd > 0) sceIoWrite(KprintfFd, buf, buf_len);
	if (RUNNING_ON_EMULATOR) {
		sceIoDevctl("emulator:", EMULATOR_DEVCTL__SEND_OUTPUT, (void *)buf, buf_len, NULL, 0);
	}

	if (buf_len < sizeof(temp)) {
		if (HAS_DISPLAY) {
			memcpy(temp, buf, buf_len);
			temp[buf_len] = 0;

			//Kprintf("%s", temp);
			pspDebugScreenPrintf("%s", temp);
		}
	}
	
	if (stdout_back._write != NULL) {
		return stdout_back._write(ptr, cookie, buf, buf_len);
	} else {
		return buf_len;
	}
}

typedef int (*SdkVerFunc)(u32 ver);
typedef struct SdkVerFuncTable {
	u32 id;
	SdkVerFunc func;
} SdkVerFuncTable;

static SdkVerFuncTable sdkVerFuncs[] = {
	{0, sceKernelSetCompiledSdkVersion},
	{370, sceKernelSetCompiledSdkVersion370},
	{380, sceKernelSetCompiledSdkVersion380_390},
	{395, sceKernelSetCompiledSdkVersion395},
	{401, sceKernelSetCompiledSdkVersion401_402},
	{500, sceKernelSetCompiledSdkVersion500_505},
	{507, sceKernelSetCompiledSdkVersion507},
	{600, sceKernelSetCompiledSdkVersion600_602},
	{603, sceKernelSetCompiledSdkVersion603_605},
	{606, sceKernelSetCompiledSdkVersion606},
};

static void updateSdkVer(int argc, char *argv[]) {
	int i = 0;
	u32 ver = 0xFFFFFFFF;
	u32 funcID = 0;
	for (i = 1; i < argc; ++i) {
		if (!strncmp(argv[i], "--sdkver=", strlen("--sdkver="))) {
			ver = strtoul(argv[i] + strlen("--sdkver="), NULL, 16);
		}
		if (!strncmp(argv[i], "--sdkver-func=", strlen("--sdkver-func="))) {
			funcID = strtol(argv[i] + strlen("--sdkver-func="), NULL, 10);
		}
	}

	SdkVerFunc func = NULL;
	for (i = 0; i < sizeof(sdkVerFuncs) / sizeof(sdkVerFuncs[0]); ++i) {
		if (sdkVerFuncs[i].id == funcID) {
			func = sdkVerFuncs[i].func;
		}
	}

	if (func == NULL) {
		fprintf(stderr, "Unknown sdkver-func value.\n");
		exit(1);
	}

	if (ver != 0xFFFFFFFF) {
		if (func(ver) != 0) {
			printf("WARNING: Setting sdkver returned failure.\n");
		}
	}
}

void test_begin() {
	if (HAS_DISPLAY) {
		pspDebugScreenInit();
	}

	if (RUNNING_ON_EMULATOR && !HAS_DISPLAY) {
		fclose(stdout);
		stdout = fmemopen(alloca(4), 4, "wb");
		stdout_back._write = NULL;
		
		//stderr = stdout;
		setbuf(stdout, NULL);
	} else {
    // Send the output to the host.
		freopen("host0:/__testoutput.txt", "wb", stdout);
		freopen("host0:/__testerror.txt", "wb", stderr);
		stdout_back._write = stdout->_write;
	}
	stdout->_write = writeStdoutHook;

	setvbuf(stdout, NULL, _IONBF, 0);
	setvbuf(stderr, NULL, _IONBF, 0);
	
	setbuf(stderr, NULL);

	reschedThread = sceKernelCreateThread("resched", &reschedFunc, sceKernelGetThreadCurrentPriority(), 0x1000, 0, NULL);
}

void test_end() {
	flushschedf();

	fflush(stdout);
	fflush(stderr);
	
	fclose(stdout);
	fclose(stderr);

	if (!RUNNING_ON_EMULATOR) {
		FILE *finish = fopen("host0:/__testfinish.txt", "wb");
		if (finish)
		{
			fwrite("1", sizeof(char), 1, finish);
			fclose(finish);
		}
	}

  // Disabled the wait, much more convienent when running automated.
	if (0 && !RUNNING_ON_EMULATOR) {
		SceCtrlData key;
		while (1) {
			sceCtrlReadBufferPositive(&key, 1);
			if (key.Buttons & PSP_CTRL_CROSS) break;
		}
	}
	
	//fclose(stdout);
	sceKernelExitGame();
	
	exit(0);
}

int test_psp_exit_callback(int arg1, int arg2, void *common) {
	exit(0);
	return 0;
}

int test_psp_callback_thread(SceSize args, void *argp) {
	int cbid;
	cbid = sceKernelCreateCallback("Exit Callback", test_psp_exit_callback, NULL);
	sceKernelRegisterExitCallback(cbid);
	sceKernelSleepThreadCB();
	return 0;
}

int test_psp_setup_callbacks(void) {
	int thid = 0;
	thid = sceKernelCreateThread("update_thread",  test_psp_callback_thread, 0x11, 0xFA0, 0, 0);
	if (thid >= 0) sceKernelStartThread(thid, 0, 0);
	return thid;
}

//#define START_WITH "ms0:/PSP/GAME/virtual"

/*
void emitInt(int v) {
	asm("syscall 0x1010");
}

void emitFloat(float v) {
	asm("syscall 0x1011");
}

void emitString(char *v) {
	asm("syscall 0x1012");
}

void emitComment(char *v) {
	asm("syscall 0x1012");
}

void emitMemoryBlock(void *address, unsigned int size) {
	asm("syscall 0x1013");
}

void emitHex(void *address, unsigned int size) {
	asm("syscall 0x1014");
}
*/

unsigned char bmpHeader[54] = {
	0x42, 0x4D, 0x38, 0x80, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x36, 0x00,
	0x00, 0x00, 0x28, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x10, 0x01,
	0x00, 0x00, 0x01, 0x00, 0x20, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x80,
	0x08, 0x00, 0x12, 0x0B, 0x00, 0x00, 0x12, 0x0B, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00
};


uint extractBits(uint value, int offset, int size) {
	return (value >> offset) & ((1 << size) - 1);
}

void emulatorEmitScreenshot() {
	int file;

	if (RUNNING_ON_EMULATOR) {
		sceIoDevctl("kemulator:", EMULATOR_DEVCTL__EMIT_SCREENSHOT, NULL, 0, NULL, 0);
	}
	else
	{
		uint topaddr;
		int bufferwidth;
		int pixelformat;

		sceDisplayGetFrameBuf((void **)&topaddr, &bufferwidth, &pixelformat, 0);
		
        if (topaddr & 0x80000000) {
            topaddr |= 0xA0000000;
        } else {
            topaddr |= 0x40000000;
        }
	
		if ((file = sceIoOpen("host0:/__screenshot.bmp", PSP_O_CREAT | PSP_O_WRONLY | PSP_O_TRUNC, 0777)) >= 0) {
			int y, x;
			uint c;
			uint* vram_row;
			uint* row_buf = (uint *)malloc(512 * 4);
			sceIoWrite(file, &bmpHeader, sizeof(bmpHeader));
			for (y = 0; y < 272; y++) {
				vram_row = (uint *)(topaddr + 512 * 4 * (271 - y));
				for (x = 0; x < 512; x++) {
					c = vram_row[x];
					/*
					row_buf[x] = (
						((extractBits(c,  0, 8)) <<  0) |
						((extractBits(c,  8, 8)) <<  8) |
						((extractBits(c, 16, 8)) << 16) |
						((                0x00 ) << 24) |
					0);
					*/
					row_buf[x] = (
						((extractBits(c, 16, 8)) <<  0) |
						((extractBits(c,  8, 8)) <<  8) |
						((extractBits(c,  0, 8)) << 16) |
						((                0x00 ) << 24) |
					0);
				}
				sceIoWrite(file, row_buf, 512 * 4);
			}
			free(row_buf);
			//sceIoWrite(file, (void *)topaddr, bufferwidth * 272 * 4);
			//sceIoFlush();
			sceIoClose(file);
		}
	}
}

void emulatorSendSceCtrlData(SceCtrlData* pad_data) {
	sceIoDevctl("kemulator:", EMULATOR_DEVCTL__SEND_CTRLDATA, pad_data, sizeof(SceCtrlData), NULL, 0);
}

int main(int argc, char *argv[]) {
	int retval = 0;

	//KprintfFd = sceIoOpen("emulator:/Kprintf", O_WRONLY, 0777);
	//RUNNING_ON_EMULATOR = (KprintfFd > 0);
	
	if (sceIoDevctl("kemulator:", EMULATOR_DEVCTL__IS_EMULATOR, NULL, 0, NULL, 0) == 0) {
		RUNNING_ON_EMULATOR = 1;
	}
	
	// TEMP HACK
	//RUNNING_ON_EMULATOR = 0;

	if (RUNNING_ON_EMULATOR) {
		sceIoDevctl("kemulator:", EMULATOR_DEVCTL__GET_HAS_DISPLAY, NULL, 0, &HAS_DISPLAY, sizeof(HAS_DISPLAY));
	}
	
	#ifdef GRAPHIC_TEST
		HAS_DISPLAY = 0;
	#endif

	//if (strncmp(argv[0], START_WITH, strlen(START_WITH)) == 0) RUNNING_ON_EMULATOR = 1;

	if (!RUNNING_ON_EMULATOR) {
		test_psp_setup_callbacks();
	}
	atexit(sceKernelExitGame);

	test_begin();
	{
		pspDebugScreenPrintf("RUNNING_ON_EMULATOR: %s - %s\n", RUNNING_ON_EMULATOR ? "yes" : "no", argv[0]);
		updateSdkVer(argc, argv);

		retval = test_main(argc, argv);
	}
	test_end();
	
	return retval;
}
