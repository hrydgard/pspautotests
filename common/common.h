#ifndef _COMMON_H
#define _COMMON_H 1

#ifdef __cplusplus
extern "C" {
#endif

// Just to make IntelliSense happy, not attempting to compile.
#ifdef _MSC_VER
#define __builtin_va_list int
#include <stdarg.h>

#define __inline__ inline
#define __attribute__(...)
#define __extension__

#ifndef __builtin_va_start
void __va_start(va_list, ...);
void *__va_arg(va_list, ...);
void __va_end(va_list);

#define __builtin_va_start __va_start
#define __builtin_va_arg __va_arg
#define __builtin_va_end __va_end
#endif
#endif

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <pspctrl.h>
#include <pspkerneltypes.h>

#undef main
#define main test_main

extern unsigned int RUNNING_ON_EMULATOR;
extern unsigned int CHECKPOINT_ENABLE_TIME;
// Causes rescheduling (sceIoWrite) but easier to debug in the emulator.
extern unsigned int CHECKPOINT_OUTPUT_DIRECT;
extern char schedfBuffer[65536];
extern unsigned int schedfBufferPos;

void schedf(const char *format, ...);
void flushschedf();
int reschedFunc(SceSize argc, void *argp);
void checkpoint(const char *format, ...);
void checkpointNext(const char *title);

#define ARRAY_SIZE(a) (sizeof((a)) / (sizeof((a)[0])))

void emulatorEmitScreenshot();
void emulatorSendSceCtrlData(SceCtrlData* pad_data);

/*
void emitInt(int v);
void emitFloat(float v);
void emitString(char *v);
void emitComment(char *v);
void emitMemoryBlock(void *address, unsigned int size);
void emitHex(void *address, unsigned int size);
#define emitStringf(format, ...) { char temp[1024]; sprintf(temp, format, __VA_ARGS__); emitString(temp); }
*/

#ifdef __cplusplus
}
#endif

#endif