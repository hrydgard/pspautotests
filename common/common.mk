# Include this to simplify your test include file.
#
# Usage:
#
# TARGETS=mytest
# include ../../path/to/common/common.mk

BUILD_PRX = 1
USE_PSPSDK_LIBC = 1
PSP_FW_VERSION = 500

INCDIR := $(INCDIR) . $(COMMON_DIR)
LIBDIR := $(LIBDIR) . $(COMMON_DIR)

ifndef CFLAGS
CFLAGS = -g -G0 -Wall -O0 -fno-strict-aliasing
endif
ifndef CXXFLAGS
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti 
endif
ifndef ASFLAGS
ASFLAGS = $(CFLAGS)
endif
ifndef LDFLAGS
LDFLAGS = -G0
endif

ifndef LIBS
LIBS = -lpspgu -lpsprtc -lpspctrl -lpspmath -lcommon -lc -lm
endif

TARGET = $(firstword $(TARGETS))
OBJS = $(firstword $(TARGETS)).o

PSPSDK = $(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.elf: %.o $(EXTRA_OBJS) $(EXPORT_OBJ)
	$(LINK.c) $^ $(LIBS) -o $@
	$(FIXUP) $@

%.prx: %.elf
	psp-prxgen $< $@

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

all: $(TARGETS:=.prx)
clean: EXTRA_TARGETS:=$(EXTRA_TARGETS) $(TARGETS:=.prx)
