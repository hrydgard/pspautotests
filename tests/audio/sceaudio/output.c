#include <common.h>
#include <pspaudio.h>
#include <psputility.h>

int channel;
int doAudioOutputBlocking(int vol, void *buf) {
	return sceAudioOutputBlocking(channel, vol, buf);
}

int doAudioOutputPannedBlocking(int vol, void *buf) {
	return sceAudioOutputPannedBlocking(channel, vol, vol, buf);
}

int doAudioRest() {
	return sceAudioGetChannelRestLength(channel);
}

const char *roughDurationDesc(u64 t) {
	if (t < 1000)
		return "very little time";
	if (t < 85000)
		return "less than 85ms";
	if (t < 86000)
		return "about the right time for 48kHz";
	if (t < 90000)
		return "less than 90ms";
	if (t < 92000)
		return "less than 92ms";
	if (t < 93000)
		return "about the right time for 44.1kHz";
	if (t < 98000)
		return "less than 98ms";
	return "a long time";
}

void testBlockingRun(const char *title, int outputFunc(int, void*), int restFunc(), int vol, u32 *data) {
	u64 startTicks, totalTicks;
	int result;

	startTicks = sceKernelGetSystemTimeWide();
	result = outputFunc(vol, data);
	totalTicks = sceKernelGetSystemTimeWide() - startTicks;
	checkpoint("  %s: 0x%x (took %s, remaining: 0x%x samples)", title, result, roughDurationDesc(totalTicks), restFunc());
}

void testBlockingType(int outputFunc(int, void*), int restFunc(), int vol, u32 *data) {
	u64 startTicks, totalTicks;
	int result;

	testBlockingRun("first", outputFunc, restFunc, vol, data);
	testBlockingRun("second", outputFunc, restFunc, vol, data);

	startTicks = sceKernelGetSystemTimeWide();
	while (sceKernelGetSystemTimeWide() - startTicks < 94000) {
		sceKernelDelayThread(2000);
	}
	checkpoint("    after delay, before output - remaining: 0x%x samples", restFunc());
	
	testBlockingRun("after delay", outputFunc, restFunc, vol, data);
	testBlockingRun("NULL", outputFunc, restFunc, vol, NULL);
	testBlockingRun("NULL", outputFunc, restFunc, vol, NULL);
	testBlockingRun("NULL", outputFunc, restFunc, vol, NULL);
}

void testBlocking() {
	u32 data[4096 + 64];
	memset(data, 0, sizeof(data));

	channel = sceAudioChReserve(PSP_AUDIO_NEXT_CHANNEL, 4096 + 64, PSP_AUDIO_FORMAT_STEREO);
	sceAudioSetChannelDataLen(channel, 4096 + 64);
	sceAudioChangeChannelVolume(channel, 0x8000, 0x8000);

	checkpointNext("sceAudioOutputBlocking:");
	testBlockingType(&doAudioOutputBlocking, &doAudioRest, -1, data);

	checkpointNext("sceAudioOutputPannedBlocking:");
	testBlockingType(&doAudioOutputPannedBlocking, &doAudioRest, 0x8000, data);
	sceAudioChRelease(channel);

	channel = sceAudioChReserve(0, 4096 + 64, PSP_AUDIO_FORMAT_STEREO);
	sceAudioSetChannelDataLen(channel, 4096 + 64);
	sceAudioChangeChannelVolume(channel, 0x8000, 0x8000);

	checkpointNext("sceAudioOutputBlocking channel 0:");
	testBlockingType(&doAudioOutputBlocking, &doAudioRest, -1, data);

	checkpointNext("sceAudioOutputPannedBlocking channel 0:");
	testBlockingType(&doAudioOutputPannedBlocking, &doAudioRest, 0x8000, data);
	sceAudioChRelease(channel);

	checkpointNext("sceAudioOutput2OutputBlocking:");
	sceAudioOutput2Reserve(4096);
	testBlockingType(&sceAudioOutput2OutputBlocking, &sceAudioOutput2GetRestSample, 0x8000, data);
	sceAudioOutput2Release();

	checkpointNext("sceAudioSRCOutputBlocking:");
	sceAudioSRCChReserve(4096, 44100, 2);
	testBlockingType(&sceAudioSRCOutputBlocking, &sceAudioOutput2GetRestSample, 0x8000, data);
	sceAudioSRCChRelease();
}

int main(int argc, char *argv[]) {
	int result;
	int i;
	u32 data[256];
	memset(data, 0, sizeof(data));

	int channel = sceAudioChReserve(PSP_AUDIO_NEXT_CHANNEL, ARRAY_SIZE(data), PSP_AUDIO_FORMAT_STEREO);
	sceAudioSetChannelDataLen(channel, ARRAY_SIZE(data));
	sceAudioChangeChannelVolume(channel, 0x8000, 0x8000);

	checkpointNext("sceAudioOutput:");
	checkpoint("Normal: %08x", sceAudioOutput(channel, 0x8000, data));
	checkpoint("Twice: %08x", sceAudioOutput(channel, 0x8000, data));
	checkpoint("Blocking: %08x", sceAudioOutputBlocking(channel, 0x8000, data));
	checkpoint("Bad channel (blocking): %08x", sceAudioOutputBlocking(-1, 0x8000, data));
	checkpoint("Unreserved channel (blocking): %08x", sceAudioOutputBlocking(5, 0x8000, data));
	checkpoint("NULL data (blocking): %08x", sceAudioOutputBlocking(channel, 0x8000, NULL));
	checkpoint("NULL data (non blocking): %08x", sceAudioOutput(channel, 0x8000, NULL));

	// NOTE: It appears (when using noise) that any negative volume means "use default" or something.
	checkpointNext("Volumes:");
	const static int volumes[] = {-0x100, -2, -1, 0, 1, 2, 0x100, 0xFFFE, 0xFFFF, 0x10000, 0x10001, 0xFFFFE, 0xFFFFF, 0x100000, 0x100001, 0x80000000, 0x7FFFFFFF};
	for (i = 0; i < ARRAY_SIZE(volumes); ++i) {
		result = sceAudioOutput(channel, volumes[i], NULL);
		checkpoint("  Non-blocking %d: %08x", volumes[i], result);
		result = sceAudioOutputBlocking(channel, volumes[i], NULL);
		checkpoint("  Blocking %d: %08x", volumes[i], result);
	}

	checkpointNext("Panned volumes:");
	for (i = 0; i < ARRAY_SIZE(volumes); ++i) {
		result = sceAudioOutputPanned(channel, volumes[i], volumes[i], NULL);
		checkpoint("  Non-blocking %d: %08x", volumes[i], result);
		result = sceAudioOutputPannedBlocking(channel, volumes[i], volumes[i], NULL);
		checkpoint("  Blocking %d: %08x", volumes[i], result);
	}

	sceAudioChRelease(channel);

	sceAudioOutput2Reserve(ARRAY_SIZE(data));

	checkpointNext("sceAudioOutput2OutputBlocking:");
	checkpoint("Normal: %08x", sceAudioOutput2OutputBlocking(0x8000, data));
	checkpoint("NULL data: %08x", sceAudioOutput2OutputBlocking(0x8000, NULL));

	checkpointNext("Volumes:");
	for (i = 0; i < ARRAY_SIZE(volumes); ++i) {
		result = sceAudioOutput2OutputBlocking(volumes[i], data);
		checkpoint("  %d: %08x", volumes[i], result);
	}

	sceAudioSRCOutputBlocking(0, NULL);
	sceAudioOutput2Release();

	sceAudioSRCChReserve(ARRAY_SIZE(data), 48000, 2);

	checkpointNext("sceAudioSRCOutputBlocking:");
	checkpoint("Normal: %08x", sceAudioSRCOutputBlocking(0x8000, data));
	checkpoint("NULL data: %08x", sceAudioSRCOutputBlocking(0x8000, NULL));

	checkpointNext("Volumes:");
	for (i = 0; i < ARRAY_SIZE(volumes); ++i) {
		result = sceAudioSRCOutputBlocking(volumes[i], data);
		checkpoint("  %d: %08x", volumes[i], result);
	}

	sceAudioSRCOutputBlocking(0, NULL);
	sceAudioSRCChRelease();

	testBlocking();

	return 0;
}
