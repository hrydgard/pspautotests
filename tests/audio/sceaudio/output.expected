[x] sceAudioOutput:
[x] Normal: 00000100
[x] Twice: 80260002
[r] Blocking: 00000100
[x] Bad channel (blocking): 80260003
[x] Unreserved channel (blocking): 80260001
[r] NULL data (blocking): 00000100
[x] NULL data (non blocking): 00000100

[x] Volumes:
[x]   Non-blocking -256: 00000100
[x]   Blocking -256: 00000100
[x]   Non-blocking -2: 00000100
[x]   Blocking -2: 00000100
[x]   Non-blocking -1: 00000100
[x]   Blocking -1: 00000100
[x]   Non-blocking 0: 00000100
[x]   Blocking 0: 00000100
[x]   Non-blocking 1: 00000100
[x]   Blocking 1: 00000100
[x]   Non-blocking 2: 00000100
[x]   Blocking 2: 00000100
[x]   Non-blocking 256: 00000100
[x]   Blocking 256: 00000100
[x]   Non-blocking 65534: 00000100
[x]   Blocking 65534: 00000100
[x]   Non-blocking 65535: 00000100
[x]   Blocking 65535: 00000100
[x]   Non-blocking 65536: 8026000b
[x]   Blocking 65536: 8026000b
[x]   Non-blocking 65537: 8026000b
[x]   Blocking 65537: 8026000b
[x]   Non-blocking 1048574: 8026000b
[x]   Blocking 1048574: 8026000b
[x]   Non-blocking 1048575: 8026000b
[x]   Blocking 1048575: 8026000b
[x]   Non-blocking 1048576: 8026000b
[x]   Blocking 1048576: 8026000b
[x]   Non-blocking 1048577: 8026000b
[x]   Blocking 1048577: 8026000b
[x]   Non-blocking -2147483648: 00000100
[x]   Blocking -2147483648: 00000100
[x]   Non-blocking 2147483647: 8026000b
[x]   Blocking 2147483647: 8026000b

[x] Panned volumes:
[x]   Non-blocking -256: 00000100
[x]   Blocking -256: 8026000b
[x]   Non-blocking -2: 00000100
[x]   Blocking -2: 8026000b
[x]   Non-blocking -1: 00000100
[x]   Blocking -1: 8026000b
[x]   Non-blocking 0: 00000100
[x]   Blocking 0: 00000100
[x]   Non-blocking 1: 00000100
[x]   Blocking 1: 00000100
[x]   Non-blocking 2: 00000100
[x]   Blocking 2: 00000100
[x]   Non-blocking 256: 00000100
[x]   Blocking 256: 00000100
[x]   Non-blocking 65534: 00000100
[x]   Blocking 65534: 00000100
[x]   Non-blocking 65535: 00000100
[x]   Blocking 65535: 00000100
[x]   Non-blocking 65536: 8026000b
[x]   Blocking 65536: 8026000b
[x]   Non-blocking 65537: 8026000b
[x]   Blocking 65537: 8026000b
[x]   Non-blocking 1048574: 8026000b
[x]   Blocking 1048574: 8026000b
[x]   Non-blocking 1048575: 8026000b
[x]   Blocking 1048575: 8026000b
[x]   Non-blocking 1048576: 8026000b
[x]   Blocking 1048576: 8026000b
[x]   Non-blocking 1048577: 8026000b
[x]   Blocking 1048577: 8026000b
[x]   Non-blocking -2147483648: 00000100
[x]   Blocking -2147483648: 8026000b
[x]   Non-blocking 2147483647: 8026000b
[x]   Blocking 2147483647: 8026000b

[x] sceAudioOutput2OutputBlocking:
[x] Normal: 00000100
[r] NULL data: 00000000

[x] Volumes:
[x]   -256: 8026000b
[x]   -2: 8026000b
[x]   -1: 8026000b
[x]   0: 00000100
[r]   1: 00000100
[r]   2: 00000100
[r]   256: 00000100
[r]   65534: 00000100
[r]   65535: 00000100
[r]   65536: 00000100
[r]   65537: 00000100
[r]   1048574: 00000100
[r]   1048575: 00000100
[x]   1048576: 8026000b
[x]   1048577: 8026000b
[x]   -2147483648: 8026000b
[x]   2147483647: 8026000b

[x] sceAudioSRCOutputBlocking:
[x] Normal: 00000100
[r] NULL data: 00000000

[x] Volumes:
[x]   -256: 8026000b
[x]   -2: 8026000b
[x]   -1: 8026000b
[x]   0: 00000100
[r]   1: 00000100
[r]   2: 00000100
[r]   256: 00000100
[r]   65534: 00000100
[r]   65535: 00000100
[r]   65536: 00000100
[r]   65537: 00000100
[r]   1048574: 00000100
[r]   1048575: 00000100
[x]   1048576: 8026000b
[x]   1048577: 8026000b
[x]   -2147483648: 8026000b
[x]   2147483647: 8026000b

[x] sceAudioOutputBlocking:
[x]   first: 0x1040 (took very little time, remaining: 0x1000 samples)
[r]   second: 0x1040 (took about the right time for 44.1kHz, remaining: 0x1040 samples)
[r]     after delay, before output - remaining: 0x0 samples
[x]   after delay: 0x1040 (took very little time, remaining: 0x1040 samples)
[r]   NULL: 0x1040 (took about the right time for 44.1kHz, remaining: 0x0 samples)
[x]   NULL: 0x1040 (took very little time, remaining: 0x0 samples)
[x]   NULL: 0x1040 (took very little time, remaining: 0x0 samples)

[x] sceAudioOutputPannedBlocking:
[x]   first: 0x1040 (took very little time, remaining: 0x1000 samples)
[r]   second: 0x1040 (took about the right time for 44.1kHz, remaining: 0x1040 samples)
[r]     after delay, before output - remaining: 0x0 samples
[x]   after delay: 0x1040 (took very little time, remaining: 0x1040 samples)
[r]   NULL: 0x1040 (took about the right time for 44.1kHz, remaining: 0x0 samples)
[x]   NULL: 0x1040 (took very little time, remaining: 0x0 samples)
[x]   NULL: 0x1040 (took very little time, remaining: 0x0 samples)

[x] sceAudioOutputBlocking channel 0:
[x]   first: 0x1040 (took very little time, remaining: 0x1000 samples)
[r]   second: 0x1040 (took about the right time for 44.1kHz, remaining: 0x1040 samples)
[r]     after delay, before output - remaining: 0x0 samples
[x]   after delay: 0x1040 (took very little time, remaining: 0x1040 samples)
[r]   NULL: 0x1040 (took about the right time for 44.1kHz, remaining: 0x0 samples)
[x]   NULL: 0x1040 (took very little time, remaining: 0x0 samples)
[x]   NULL: 0x1040 (took very little time, remaining: 0x0 samples)

[x] sceAudioOutputPannedBlocking channel 0:
[x]   first: 0x1040 (took very little time, remaining: 0x1000 samples)
[r]   second: 0x1040 (took about the right time for 44.1kHz, remaining: 0x1040 samples)
[r]     after delay, before output - remaining: 0x0 samples
[x]   after delay: 0x1040 (took very little time, remaining: 0x1000 samples)
[r]   NULL: 0x1040 (took about the right time for 44.1kHz, remaining: 0x0 samples)
[x]   NULL: 0x1040 (took very little time, remaining: 0x0 samples)
[x]   NULL: 0x1040 (took very little time, remaining: 0x0 samples)

[x] sceAudioOutput2OutputBlocking:
[x]   first: 0x1000 (took very little time, remaining: 0x1000 samples)
[r]   second: 0x1000 (took about the right time for 44.1kHz, remaining: 0x1000 samples)
[r]     after delay, before output - remaining: 0x0 samples
[x]   after delay: 0x1000 (took very little time, remaining: 0x1000 samples)
[r]   NULL: 0x0 (took about the right time for 44.1kHz, remaining: 0x0 samples)
[r]   NULL: 0x0 (took very little time, remaining: 0x0 samples)
[x]   NULL: 0x0 (took very little time, remaining: 0x0 samples)

[x] sceAudioSRCOutputBlocking:
[x]   first: 0x1000 (took very little time, remaining: 0x1000 samples)
[r]   second: 0x1000 (took about the right time for 44.1kHz, remaining: 0x1000 samples)
[r]     after delay, before output - remaining: 0x0 samples
[x]   after delay: 0x1000 (took very little time, remaining: 0x1000 samples)
[r]   NULL: 0x0 (took about the right time for 44.1kHz, remaining: 0x0 samples)
[r]   NULL: 0x0 (took very little time, remaining: 0x0 samples)
[x]   NULL: 0x0 (took very little time, remaining: 0x0 samples)
